File: distributed-job-manager/src/main/java/ru/fix/distributed/job/manager/strategy/EvenlySpreadAssignmentStrategy.java
package ru.fix.distributed.job.manager.strategy;

import ru.fix.distributed.job.manager.model.AssignmentState;
import ru.fix.distributed.job.manager.model.JobId;
import ru.fix.distributed.job.manager.model.WorkItem;
import ru.fix.distributed.job.manager.model.WorkerId;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class EvenlySpreadAssignmentStrategy extends AbstractAssignmentStrategy {

    @Override
    public AssignmentState reassignAndBalance(
            Map<JobId, Set<WorkerId>> availability,
            AssignmentState prevAssignment,
            AssignmentState currentAssignment,
            Set<WorkItem> itemsToAssign
    ) {
        for (Map.Entry<JobId, Set<WorkerId>> jobEntry : availability.entrySet()) {
            JobId jobId = jobEntry.getKey();
            Set<WorkerId> availableWorkers = jobEntry.getValue();

            Set<WorkItem> itemsToAssignForJob = getWorkItemsByJob(jobId, itemsToAssign);
            availableWorkers.forEach(e -> currentAssignment.putIfAbsent(e, new HashSet<>()));

            int workersCount = availableWorkers.size();
            int workItemsCount = itemsToAssignForJob.size();
            int limitWorkItemsOnWorker = limitWorkItemsOnWorker(workItemsCount, workersCount);
            int itemsAssignedFromPreviousCounter = 0;

            for (WorkItem item : itemsToAssignForJob) {
                WorkerId workerFromPrevious = prevAssignment.getWorkerOfWorkItem(item);

                if (prevAssignment.containsWorkItem(item)
                        && currentAssignment.containsKey(workerFromPrevious)
                        && currentAssignment.isBalanced()
                        && currentAssignment.isBalancedByJobId(item.getJobId())
                        && itemsAssignedFromPreviousCounter < limitWorkItemsOnWorker
                ) {
                    currentAssignment.addWorkItem(workerFromPrevious, item);
                    itemsAssignedFromPreviousCounter++;

                    int limitItemsOnWorkerInPreviousState = prevAssignment
                            .get(workerFromPrevious, jobId).size();
                    if (limitItemsOnWorkerInPreviousState <= itemsAssignedFromPreviousCounter) {
                        itemsAssignedFromPreviousCounter = 0;
                    }
                } else {
                    WorkerId lessBusyWorker = currentAssignment
                            .getLessBusyWorkerFromAvailableWorkers(availableWorkers);
                    currentAssignment.addWorkItem(lessBusyWorker, item);
                }
                itemsToAssign.remove(item);
            }
        }
        return currentAssignment;
    }

    private int limitWorkItemsOnWorker(int itemsCount, int workersCount) {
        return itemsCount / workersCount + ((itemsCount % workersCount == 0) ? 0 : 1);
    }

}

File: distributed-job-manager/src/main/kotlin/ru/fix/distributed/job/manager/strategy/EvenlySpreadAssignmentStrategy.kt
package ru.fix.distributed.job.manager.strategy

import ru.fix.distributed.job.manager.model.AssignmentState
import ru.fix.distributed.job.manager.model.JobId
import ru.fix.distributed.job.manager.model.WorkItem
import ru.fix.distributed.job.manager.model.WorkerId
import java.util.*

class EvenlySpreadAssignmentStrategy : AbstractAssignmentStrategy() {

    override fun reassignAndBalance(
            availability: MutableMap<JobId, MutableSet<WorkerId>>,
            prevAssignment: AssignmentState,
            currentAssignment: AssignmentState,
            itemsToAssign: MutableSet<WorkItem>
    ) {
        for ((jobId, availableWorkers) in availability) {
            val itemsToAssignForJob = getWorkItemsByJob(jobId, itemsToAssign)
            availableWorkers.forEach { currentAssignment.putIfAbsent(it, HashSet<WorkItem>()) }

            for (item in itemsToAssignForJob) {
                val workerFromPrevious = prevAssignment.getWorkerOfWorkItem(item)

                if (prevAssignment.containsWorkItem(item)
                        && availableWorkers.contains(workerFromPrevious)) {
                    currentAssignment.addWorkItem(workerFromPrevious, item)
                } else {
                    val lessBusyWorker = currentAssignment
                            .getLessBusyWorkerWithJobId(jobId, availableWorkers)
                    currentAssignment.addWorkItem(lessBusyWorker, item)
                }
                itemsToAssign.remove(item)
            }

            while (!currentAssignment.isBalancedByJobId(jobId, availableWorkers)) {
                val mostBusyWorker = currentAssignment
                        .getMostBusyWorkerWithJobId(jobId, availableWorkers)
                val lessBusyWorker = currentAssignment
                        .getLessBusyWorkerWithJobId(jobId, availableWorkers)
                val itemToMove = currentAssignment.getWorkItems(mostBusyWorker, jobId).last()
                currentAssignment.moveWorkItem(itemToMove, mostBusyWorker, lessBusyWorker)
            }
        }
    }
}
